File: backend/app/models/equipment.py
from datetime import date, datetime
from decimal import Decimal
from typing import TYPE_CHECKING

from sqlalchemy import Boolean, Date, DateTime, ForeignKey, Integer, Numeric, String, Text
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.database import Base
from app.models.base import TimestampMixin, utcnow

if TYPE_CHECKING:
    from app.models.user import User


class Equipment(Base, TimestampMixin):
    __tablename__ = "equipment"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    asset_no: Mapped[str] = mapped_column(String(50), unique=True, index=True, nullable=False)
    name: Mapped[str] = mapped_column(String(200), index=True, nullable=False)
    category: Mapped[str] = mapped_column(String(50), index=True, nullable=False)
    manufacturer: Mapped[str | None] = mapped_column(String(200), nullable=True)
    model: Mapped[str | None] = mapped_column(String(200), nullable=True)
    serial_no: Mapped[str | None] = mapped_column(String(100), nullable=True)

    location: Mapped[str | None] = mapped_column(String(200), nullable=True)
    manager_id: Mapped[int | None] = mapped_column(
        ForeignKey("users.id", ondelete="SET NULL"), index=True, nullable=True
    )

    purchase_date: Mapped[date | None] = mapped_column(Date, nullable=True)
    purchase_price: Mapped[Decimal | None] = mapped_column(Numeric(12, 2), nullable=True)
    warranty_end_date: Mapped[date | None] = mapped_column(Date, nullable=True)

    status: Mapped[str] = mapped_column(String(20), default="available", index=True, nullable=False)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    manual_url: Mapped[str | None] = mapped_column(String(500), nullable=True)
    photo_url: Mapped[str | None] = mapped_column(String(500), nullable=True)

    booking_required: Mapped[bool] = mapped_column(Boolean, default=False, nullable=False)
    is_active: Mapped[bool] = mapped_column(Boolean, default=True, nullable=False)

    deleted_at: Mapped[datetime | None] = mapped_column(DateTime, nullable=True, index=True)


class EquipmentBooking(Base, TimestampMixin):
    __tablename__ = "equipment_bookings"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    equipment_id: Mapped[int] = mapped_column(
        ForeignKey("equipment.id", ondelete="CASCADE"), index=True, nullable=False
    )
    user_id: Mapped[int] = mapped_column(
        ForeignKey("users.id", ondelete="CASCADE"), index=True, nullable=False
    )
    project_id: Mapped[int | None] = mapped_column(
        ForeignKey("projects.id", ondelete="SET NULL"), index=True, nullable=True
    )

    start_time: Mapped[datetime] = mapped_column(DateTime, index=True, nullable=False)
    end_time: Mapped[datetime] = mapped_column(DateTime, index=True, nullable=False)

    purpose: Mapped[str | None] = mapped_column(Text, nullable=True)
    status: Mapped[str] = mapped_column(String(20), default="pending", index=True, nullable=False)

    approved_by: Mapped[int | None] = mapped_column(
        ForeignKey("users.id", ondelete="SET NULL"), nullable=True
    )
    approved_at: Mapped[datetime | None] = mapped_column(DateTime, nullable=True)

    user: Mapped["User"] = relationship(foreign_keys=[user_id])


class EquipmentBorrow(Base, TimestampMixin):
    __tablename__ = "equipment_borrows"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    equipment_id: Mapped[int] = mapped_column(
        ForeignKey("equipment.id", ondelete="CASCADE"), index=True, nullable=False
    )
    borrower_id: Mapped[int] = mapped_column(
        ForeignKey("users.id", ondelete="CASCADE"), index=True, nullable=False
    )

    borrow_time: Mapped[datetime | None] = mapped_column(DateTime, nullable=True)
    expected_return_time: Mapped[datetime | None] = mapped_column(DateTime, nullable=True)
    actual_return_time: Mapped[datetime | None] = mapped_column(DateTime, nullable=True)

    purpose: Mapped[str | None] = mapped_column(Text, nullable=True)
    status: Mapped[str] = mapped_column(String(20), default="borrowed", index=True, nullable=False)
    note: Mapped[str | None] = mapped_column(Text, nullable=True)

    approved_by: Mapped[int | None] = mapped_column(
        ForeignKey("users.id", ondelete="SET NULL"), nullable=True
    )


class EquipmentMaintenance(Base, TimestampMixin):
    __tablename__ = "equipment_maintenance"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    equipment_id: Mapped[int] = mapped_column(
        ForeignKey("equipment.id", ondelete="CASCADE"), index=True, nullable=False
    )
    reporter_id: Mapped[int | None] = mapped_column(
        ForeignKey("users.id", ondelete="SET NULL"), nullable=True
    )

    type: Mapped[str] = mapped_column(String(20), default="fault", index=True, nullable=False)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)

    reported_at: Mapped[datetime | None] = mapped_column(DateTime, nullable=True)
    started_at: Mapped[datetime | None] = mapped_column(DateTime, nullable=True)
    finished_at: Mapped[datetime | None] = mapped_column(DateTime, nullable=True)

    status: Mapped[str] = mapped_column(String(20), default="reported", index=True, nullable=False)
    cost: Mapped[Decimal | None] = mapped_column(Numeric(12, 2), nullable=True)
    vendor: Mapped[str | None] = mapped_column(String(200), nullable=True)
    result: Mapped[str | None] = mapped_column(Text, nullable=True)

File: frontend/src/api/equipment.ts
import client from './client'

export interface Equipment {
  id: number
  asset_no: string
  name: string
  category: string
  manufacturer: string | null
  model: string | null
  serial_no: string | null
  location: string | null
  manager_id: number | null
  manager_name?: string | null
  status: string
  description: string | null
  manual_url: string | null
  booking_required: boolean
  is_active: boolean
  qr_token?: string | null
  qr_url?: string | null
  current_booking?: { id: number; start_time: string; end_time: string; user_name: string | null } | null
  next_booking?: { id: number; start_time: string; end_time: string; user_name: string | null } | null
  latest_maintenance?: { id: number; type: string; status: string; reported_at: string | null } | null
}

export interface EquipmentBooking {
  id: number
  equipment_id: number
  equipment_name?: string
  user_id: number
  user_name?: string | null
  project_id: number | null
  start_time: string
  end_time: string
  purpose: string | null
  status: string
}

export interface EquipmentBorrow {
  id: number
  equipment_id: number
  equipment_name?: string
  borrower_id: number
  borrower_name?: string | null
  borrow_time: string | null
  expected_return_time: string | null
  actual_return_time: string | null
  purpose: string | null
  status: string
  note: string | null
  is_overdue?: boolean
}

export interface EquipmentMaintenance {
  id: number
  equipment_id: number
  equipment_name?: string
  reporter_id: number | null
  reporter_name?: string | null
  type: string
  description: string | null
  reported_at: string | null
  started_at: string | null
  finished_at: string | null
  status: string
  cost: string | null
  vendor: string | null
  result: string | null
}

export function listEquipment(params: Record<string, unknown> = {}) {
  return client.get<{ data: { items: Equipment[]; total: number } }>('/equipment', { params })
}

export function getEquipment(id: number) {
  return client.get<{ data: Equipment }>(`/equipment/${id}`)
}

export function createEquipment(payload: Record<string, unknown>) {
  return client.post('/equipment', payload)
}

export function updateEquipment(id: number, payload: Record<string, unknown>) {
  return client.patch(`/equipment/${id}`, payload)
}

export function listBookings(params: Record<string, unknown> = {}) {
  return client.get<{ data: { items: EquipmentBooking[]; total: number } }>('/equipment-bookings', { params })
}

export function createBooking(payload: Record<string, unknown>) {
  return client.post('/equipment-bookings', payload)
}

export function cancelBooking(id: number) {
  return client.post(`/equipment-bookings/${id}/cancel`)
}

export function listBorrows(params: Record<string, unknown> = {}) {
  return client.get<{ data: { items: EquipmentBorrow[]; total: number } }>('/equipment-borrows', { params })
}

export function createBorrow(payload: Record<string, unknown>) {
  return client.post('/equipment-borrows', payload)
}

export function extendBorrow(id: number, payload: { expected_return_time: string; note?: string }) {
  return client.patch(`/equipment-borrows/${id}`, payload)
}

export function returnBorrow(id: number) {
  return client.post(`/equipment-borrows/${id}/return`)
}

export function generateQr(id: number, regenerate = false) {
  return client.post<{ data: { qr_token: string; qr_url: string | null } }>(
    `/equipment/${id}/qr`,
    null,
    { params: regenerate ? { regenerate: true } : {} },
  )
}

export function resolveQr(token: string) {
  return client.get<{
    data: { resource_type: string; equipment_id: number; asset_no: string; name: string; status: string }
  }>(`/qr/${token}`)
}

export function listMaintenance(params: Record<string, unknown> = {}) {
  return client.get<{ data: { items: EquipmentMaintenance[]; total: number } }>('/equipment-maintenance', { params })
}

export function createMaintenance(payload: Record<string, unknown>) {
  return client.post('/equipment-maintenance', payload)
}

export function updateMaintenance(id: number, payload: Record<string, unknown>) {
  return client.patch(`/equipment-maintenance/${id}`, payload)
}
